File: Subway.java
import java.io.*;
import java.net.*;
import java.util.*;
import java.util.Scanner;
import java.io.*;
public class Main 
{
	public static int getHash(String td)
	{
		int lp=0;
		int mm=0;
		for (int i=0; i<td.length(); i++)
		{
			if (td.charAt(i) == ' ')
				lp=i;
			if (td.charAt(i) == 'R')
				mm=mm<<1;
			if (td.charAt(i) == 'L')
				mm=mm<<1|1;
		}
		mm=mm*100+lp;
		return mm;
	}
	public static ArrayList<Integer> solve(String st)
	{
		ArrayList<Integer> ans=new ArrayList<Integer>();
		ArrayList<Node> mem=new ArrayList<Node>();
		Set<Integer> s = new HashSet<Integer>();
		Queue<Integer> q = new LinkedList<Integer>(); 
		int l=st.length();
		int cntR=0,cntL=0,sp=0;
		for (int i=0; i<l; i++)
		{
			if (st.charAt(i) == 'R')
				cntR++;
			else if (st.charAt(i) == 'L')
				cntL++;
			else
				sp=i;
		}
		String tar="";
		for (int i=0; i<cntR; i++)
			tar=tar+"R";
		tar=tar+" ";
		for (int i=0; i<cntL; i++)
			tar=tar+"L";
		q.clear();
		s.clear();
		mem.clear();
		int ts=0,te=0;
		while (st.charAt(ts) == 'R')
			ts++;
		te=l-1;
		while (st.charAt(te) == 'L')
			te--;
		mem.add(new Node(st,sp,-1,ts,te));
		q.offer(0);
		while (!q.isEmpty())
		{
			int t=q.poll();
			int pa=t;
			sp=mem.get(t).sp;
			
			String data=mem.get(t).data;
			ts=mem.get(t).ts;
			te=mem.get(t).te;
			if (data.compareTo(tar) == 0)
			{
				ans.clear();
				ans.add(mem.get(t).sp+1);
				t=mem.get(t).pa;
				while (true)
				{
					ans.add(mem.get(t).sp+1);
					if (mem.get(t).pa == -1)
							break;
					t=mem.get(t).pa;
				}
			}
			if (sp != 0)
			{
				String td="";
				for (int i=0; i<sp-1; i++)
					td=td+data.charAt(i);
				td=td+" "+data.charAt(sp-1);
				for (int i=sp+1; i<l; i++)
					td=td+data.charAt(i);
				
				sp=sp-1;
				int mm=getHash(td);
				if (s.contains(mm) == false)
				{
					s.add(mm);
					int tss=0,tee=0;
					while (data.charAt(tss) == 'R')
						tss++;
					tee=l-1;
					while (data.charAt(tee) == 'L')
						tee--;
					mem.add(new Node(td,sp,pa,tss,tee));
					q.offer(mem.size()-1);
				}
				sp=sp+1;
			}
			if (sp != l-1)
			{
				String td="";
				for (int i=0; i<sp; i++)
					td=td+data.charAt(i);
				td=td+data.charAt(sp+1)+" ";
				for (int i=sp+2; i<l; i++)
					td=td+data.charAt(i);
				sp=sp+1;
				int mm=getHash(td);
				if (s.contains(mm) == false)
				{
					s.add(mm);
					int tss=0,tee=0;
					while (data.charAt(tss) == 'R')
						tss++;
					tee=l-1;
					while (data.charAt(tee) == 'L')
						tee--;
					mem.add(new Node(td,sp,pa,tss,tee));
					q.offer(mem.size()-1);
				}
				sp=sp-1;
			}
			
			if (sp < l-2)
			{
				String td="";
				for (int i=0; i<sp; i++)
					td=td+data.charAt(i);
				td=td+data.charAt(sp+2)+data.charAt(sp+1)+" ";
				for (int i=sp+3; i<l; i++)
					td=td+data.charAt(i);
				sp=sp+2;
				int mm=getHash(td);
				if (s.contains(mm) == false)
				{
					s.add(mm);
					int tss=0,tee=0;
					while (data.charAt(tss) == 'R')
						tss++;
					tee=l-1;
					while (data.charAt(tee) == 'L')
						tee--;
					mem.add(new Node(td,sp,pa,tss,tee));
					q.offer(mem.size()-1);
				}
				sp=sp-2;
			}
			if (sp > 1)
			{
				String td="";
				for (int i=0; i<sp-2; i++)
					td=td+data.charAt(i);
				td=td+" "+data.charAt(sp-1)+data.charAt(sp-2);
				for (int i=sp+1; i<l; i++)
					td=td+data.charAt(i);
				sp=sp-2;
				int mm=getHash(td);
				if (s.contains(mm) == false)
				{
					s.add(mm);
					int tss=0,tee=0;
					while (data.charAt(tss) == 'R')
						tss++;
					tee=l-1;
					while (data.charAt(tee) == 'L')
						tee--;
					mem.add(new Node(td,sp,pa,tss,tee));
					q.offer(mem.size()-1);
				}
				sp=sp+2;
			}
		}
		return ans;	
	}
	public static void main(String[] args) throws UnknownHostException, IOException 
	{
		String mui="1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
		int lm=mui.length();
		String host = "218.2.197.242";  //要连接的服务端IP地址  
		int port = 6000;   //要连接的服务端对应的监听端口  
		//与服务端建立连接  
		Socket client = new Socket(host, port);  
		//建立连接后就可以往服务端写数据了  

		BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
		PrintWriter printWriter =new PrintWriter(client.getOutputStream(),true);

		System.out.println("1."+in.readLine()); 
		System.out.println("2."+in.readLine()); 
		System.out.println("3."+in.readLine()); 
		String ts=in.readLine();
		String[] tmp=ts.split("\"");
		String sha=new String("");
		for (int i=0; i<lm; i++)
			for (int j=0; j<lm; j++)
				for (int k=0; k<lm; k++)
					for (int l=0; l<lm; l++)
					{
						if (EncryptUtil.encrypt(tmp[1]+mui.charAt(i)+mui.charAt(j)+mui.charAt(k)+mui.charAt(l), "SHA-1").compareTo(tmp[3]) == 0)
						{
							System.out.println(EncryptUtil.encrypt(tmp[1]+mui.charAt(i)+mui.charAt(j)+mui.charAt(k)+mui.charAt(l), "SHA-1")+" "+tmp[3]);
							sha=sha+mui.charAt(i)+mui.charAt(j)+mui.charAt(k)+mui.charAt(l);
							k=lm;
							j=lm;
							i=lm;
							break;
						}
					}
		printWriter.println(sha);
		printWriter.flush();
		
		System.out.println("4."+in.readLine()+" "+sha); 
		System.out.println("5."+in.readLine()); 
		while (true)
		{
			System.out.println("6."+in.readLine()); 
			System.out.println("7."+in.readLine()); 
			String st=in.readLine();
			System.out.println("8."+st); //!!!!
			ArrayList<Integer> al=solve(st);
			for (int i=al.size()-2; i>=0; i--)
			{
				printWriter.println(al.get(i).toString());
				printWriter.flush();
				System.out.println("res."+in.readLine());
			}
		}
		//client.close();  
	}
}
class Node
{
	public String data;
	public Node() {
	}
	public Node(String data,int sp,int pa,int ts,int te) {
		this.data=data;
		this.sp=sp;
		this.pa=pa;
		this.ts=ts;
		this.te=te;
		
	}
	public int sp;
	public int pa;
	public int ts;
	public int te;
}

File: Subway.py
import socket
import re
import hashlib
import Queue
lis="1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
global mem
def check(data):
	a="L"
	for i in data:
		if i == ' ':
			a='R'
		if i != a:
			return False
	return True
def backtrack(t):
	t=1
def getHash(td):
	lp=0
	mm=0
	for i in range(0,len(td)):
		if td[i] == ' ':
			lp=i
		if td[i] == 'R':
			mm=mm<<1
		if td[i] == 'L':
			mm=mm<<1|1
	mm=mm*100+lp
	return mm
def solve(data):
	hs=[]
	ans=[]
	mem=[]
	tmem=[]
	cntR=0
	cntL=0
	sp=0
	l=len(data)
	for i in range(0,l):		
		if data[i] == 'R':
			cntR=cntR+1
		elif data[i] == 'L':
			cntL=cntL+1
		else:
			sp=i
	tar=""	
	for i in range(0,cntR):
		tar=tar+'R'
	tar=tar+" "
	for i in range(0,cntL):
		tar=tar+'L'
	
	q=Queue.Queue()
	ts=0
	while data[ts] == 'R':
		ts=ts+1
	te=l-1
	while data[te] == 'L':
		te=te-1
	mem.append((data,sp,-1,ts,te))
	q.put(0)
	
	while q.empty() == False:
		t=q.get()
		pa=t
		sp=mem[t][1]
		data=mem[t][0]
		ts=mem[t][3]
		te=mem[t][4]
		#print data,l,t
		#raw_input()
		if cmp(data,tar) == 0:
			ans.append(mem[t][1]+1)
			t=mem[t][2]
			while True:
				ans.append(mem[t][1]+1)
				if mem[t][2] == -1:
					break				
				t=mem[t][2]
			break
		if sp != ts:
			td=""
			for i in range(0,sp-1):
				td=td+data[i]
			td=td+' '+data[sp-1]
			for i in range(sp+1,l):
				td=td+data[i]
			sp=sp-1
			mm=getHash(td)
			if mm not in tmem:
				tmem.append(mm)
				tss=0
				while td[tss] == 'R':
					tss=tss+1
				tee=l-1
				while td[tee] == 'L':
					tee=tee-1
				mem.append((td,sp,pa,tss,tee))
				q.put(len(mem)-1)	
			sp=sp+1
		if sp != te:
			td=""
			for i in range(0,sp):
				td=td+data[i]		
			td=td+data[sp+1]+' '
			for i in range(sp+2,l):
				td=td+data[i]
			sp=sp+1
			mm=getHash(td)
			if mm not in tmem:
				tmem.append(mm)
				tss=0
				while td[tss] == 'R':
					tss=tss+1
				tee=l-1
				while td[tee] == 'L':
					tee=tee-1
				mem.append((td,sp,pa,tss,tee))
				q.put(len(mem)-1)
			sp=sp-1
		if sp < te-1:
			td=""
			for i in range(0,sp):
				td=td+data[i]
			td=td+data[sp+2]+data[sp+1]+' '
			for i in range(sp+3,l):
				td=td+data[i]
			sp=sp+2
			mm=getHash(td)
			if mm not in tmem:
				tmem.append(mm)
				tss=0
				while td[tss] == 'R':
					tss=tss+1
				tee=l-1
				while td[tee] == 'L':
					tee=tee-1
				mem.append((td,sp,pa,tss,tee))
				q.put(len(mem)-1)
			sp=sp-2
		if sp > ts+1:
			td=""
			for i in range(0,sp-2):
				td=td+data[i]
			td=td+' '+data[sp-1]+data[sp-2]
			for i in range(sp+1,l):
				td=td+data[i]
			sp=sp-2
			
			mm=getHash(td)
			if mm not in tmem:
				tmem.append(mm)
				tss=0
				while td[tss] == 'R':
					tss=tss+1
				tee=l-1
				while td[tee] == 'L':
					tee=tee-1
				mem.append((td,sp,pa,tss,tee))
				q.put(len(mem)-1)
			sp=sp+2
	print ans
	return ans
"""
ditc={}
def dfs(t,sp,dt):
	if t == 15 and sp == True:
		print dt
		ditc[dt]=solve(dt)
		return 0
	elif t == 15 and sp == False:
		return 0
	if sp == False:
		d=dt[:]
		dfs(t+1,True,d+' ')
		dfs(t+1,False,d+'R')
		dfs(t+1,False,d+'L')
	else:
		d=dt[:]
		dfs(t+1,True,d+'R')
		dfs(t+1,True,d+'L')
dfs(0,False,"")

for (k,v) in  ditc.items(): 
	print k
	print v
"""
print hashlib.sha1("GlaqzghhRHOtOzQESCzi").hexdigest()
print "bc5b62455241161bcee24786f0e4bbe5d9d5e423"
address = ('218.2.197.243', 6000)
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(address)

data = s.recv(512)
data = s.recv(512)
m=data.split('"')
print m[1],m[3]
for i in lis:
	for j in lis:
		for k in lis:
			for l in lis:
				if hashlib.sha1(m[1]+i+j+k+l).hexdigest() == m[3]:
					s.send("%c%c%c%c\n" % (i,j,k,l));
					data = s.recv(512)
					print data
					while True:
						data = s.recv(512)
						
						data=data.split('\n')
						print data,len(data)
						if len(data) == 3:
							sd=solve(data[1])
						else:
							sd=solve(data[2])
						for o in range(len(sd)-2,-1,-1):
							s.send(str(sd[o])+"\n")
							print sd[o]
							print s.recv(512)
												

s.close()
